tidy3d/plugins/metrics/base.py

```diff
@@ -0,0 +1,186 @@
+from __future__ import annotations
+
+from abc import ABC, abstractmethod
+from typing import TYPE_CHECKING, Any
+
+from tidy3d.components.base import Tidy3dBaseModel
+from tidy3d.components.types import TYPE_TAG_STR
+
+from .types import ExpressionType, NumberOrExpression, NumberType
+
+if TYPE_CHECKING:
+    from .operators import (
+        Abs,
+        Add,
+        Divide,
+        FloorDivide,
+        MatMul,
+        Modulus,
+        Multiply,
+        Negate,
+        Power,
+        Subtract,
+    )
+
+TYPE_TO_CLASS_MAP: dict[str, Any] = {}
+
+
+class Expression(Tidy3dBaseModel, ABC):
+    """
+    Base class for all expressions in the metrics module.
+
+    This class serves as the foundation for all other components in the metrics module.
+    It provides common functionality and operator overloading for derived classes.
+    """
+
+    class Config:
+        smart_union = True
+
+    @abstractmethod
+    def evaluate(self, *args: Any, **kwargs: Any) -> NumberType:
+        pass
+
+    def __call__(self, *args: Any, **kwargs: Any) -> NumberType:
+        return self.evaluate(*args, **kwargs)
+
+    def __init_subclass__(cls, **kwargs: dict[str, Any]) -> None:
+        super().__init_subclass__(**kwargs)
+        type_value = cls.__fields__.get(TYPE_TAG_STR)
+        if type_value and type_value.default:
+            TYPE_TO_CLASS_MAP[type_value.default] = cls
+
+    @classmethod
+    def parse_obj(cls, obj: dict[str, Any]) -> ExpressionType:
+        if not isinstance(obj, dict):
+            raise TypeError("Input must be a dict")
+        type_value = obj.get(TYPE_TAG_STR)
+        if type_value is None:
+            raise ValueError('Missing "type" in data')
+        subclass = TYPE_TO_CLASS_MAP.get(type_value)
+        if subclass is None:
+            raise ValueError(f"Unknown type: {type_value}")
+        return subclass(**obj)
+
+    @staticmethod
+    def _to_expression(other: NumberOrExpression | dict[str, Any]) -> ExpressionType:
+        if isinstance(other, Expression):
+            return other
+        elif isinstance(other, dict):
+            return Expression.parse_obj(other)
+        else:
+            from .variables import Constant
+
+            return Constant(other)
+
+    def __neg__(self) -> Negate:
+        from .operators import Negate
+
+        return Negate(operand=self)
+
+    def __add__(self, other: NumberOrExpression) -> Add:
+        from .operators import Add
```

Comment:
is there a reason of importing this within each method? instead of just
```py
from . import operators

...
 
    operators.Add
```

Reply:
circular imports 😄
an `Operator` is an `Expression`, and `Expression`s implement dunder methods in terms of `Operator`s

Reply:
hmm. does this tie into the forward reference stuff too?

Reply:
no these are separate things. this here is a runtime circular dependency

Reply:
hm, so the way I understand things, the Expression is a graph and Variables / Operators are the nodes of the graph. A single constant is both a Variable and a Graph, but it seems like making them all expressions leads to some circular dependencies?